### ISSUE TITLE
[ENHANCEMENT] Auto-cleanup container images that are no longer used by the Harvester system services after upgrades

### ISSUE THREAD

**Issue by starbops** [harvester/harvester#4425]
**Is your enhancement related to a problem? Please describe.**

We already have a [script](https://github.com/harvester/upgrade-helpers/blob/d490e51c76b107c8d5df2df71d0e9b7000b392d6/bin/harv-purge-images.sh) that does image clean-up for unused Harvester images, typically the container images from the previous release of Harvester. The script requires users to execute it on each Harvester node with administrator privilege. We want the experience to be less cluttered, so the tedious image clean-up task should automatically be done with the upgrade controller.

**Describe the solution you'd like**

- Such a mechanism should only be applied after a successful upgrade
- Best-effort design: log the incidents if the to-be-removed images do not exist on the nodes or if something wrong happened. It should never break the upgrade process.
- Air-gapped environment support: it shouldn't depend on the diff of the image lists on the Internet like [what we did in the past](https://github.com/harvester/upgrade-helpers/blob/d490e51c76b107c8d5df2df71d0e9b7000b392d6/bin/harv-purge-images.sh#L35C16-L35C16).

**Describe alternatives you've considered**

Unpin the previous version of the container images after a successful upgrade. Let the image GC manager do its job. But this can only happen when #4104 is delivered.

**Additional context**

With #4104, it's possible that we **pin** some Harvester-critical container images to prevent them from being GC'ed. With that said, if we don't remove those images, the available disk space will be much more limited after the upgrade over and over again.

**Comment by starbops** [comment]
The proposed solution includes two parts:

- The packaging script (harvester/harvester-installer#637): We added snippets to download all the previous versions' published image lists and pack them into the ISO image.
- The upgrade controller (#4995): At the end of an upgrade, extend the repo VM's life by adding a step. The controller will try to get the image lists of the current version (after the upgrade) and the previous version (before the upgrade) and create a Plan to purge the container images on each node according to the diff list. After the Plan is completed, the regular cleanup procedure continues.

**Comment by starbops** [comment]
**Test plan**

*Building ISO image*

Because the proposed solution introduces changes in two repositories, generating a custom-built ISO image will be a little hassle.

1. Clone the `harvester/harvester` repo and checkout to the PR branch
2. Update the `scripts/build-iso` file and make it clone the `harvester/harvester-installer` repo from the PR branch, for example:

    ```
    ...
    HARVESTER_INSTALLER_VERSION=enh-4425

    git clone --branch ${HARVESTER_INSTALLER_VERSION} --single-branch --depth 1 https://github.com/starbops/harvester-installer.git ../harvester-installer
    ...
    ```

3. Build the ISO image

    ```
    export REPO=<your-repo>
    export PUSH=true
    make && make build-iso
    ```

*Upgrade*

1. Prepare a Harvester cluster of version 1.2.1
2. Start the upgrade
3. After the upgrade ended successfully, log in to each node and check if the container images used in v1.2.1 only do not exist anymore. You can reference https://releases.rancher.com/harvester/v1.2.1/image-lists.tar.gz.

**Comment by TachunLin** [comment]
Verified **fixed** on `master-70c6cca7-head` (24/01/30). 

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ Ensure the images that are going to be removed after upgrade no longer exists   $~~$</summary>

1. From v1.2.1 to v1.3.0 master, we have the following images going to be removed: 
```
image-lists/
image-lists/harvester-repo-images-master.txt
image-lists/rke2-images-multus.linux-amd64-v1.26.11-rke2r1.txt
image-lists/rke2-images.linux-amd64-v1.26.11-rke2r1.txt
image-lists/harvester-extra-master.txt
image-lists/rancher-images-v2.7.9.txt
image-lists/harvester-images-master.txt
image-lists/rancherd-bootstrap-images-master.txt
+------------------------------------------------------+
| Images that are going to be REMOVED are listed BELOW |
+------------------------------------------------------+
docker.io/library/alpine:3
docker.io/library/busybox:1.32.0
docker.io/longhornio/backing-image-manager:v1.4.3
docker.io/longhornio/csi-attacher:v3.4.0
docker.io/longhornio/csi-node-driver-registrar:v2.5.0
docker.io/longhornio/csi-provisioner:v2.1.2
docker.io/longhornio/csi-resizer:v1.3.0
docker.io/longhornio/csi-snapshotter:v5.0.1
docker.io/longhornio/livenessprobe:v2.8.0
docker.io/longhornio/longhorn-engine:v1.4.3
docker.io/longhornio/longhorn-instance-manager:v1.4.3
docker.io/longhornio/longhorn-manager:v1.4.3
docker.io/longhornio/longhorn-share-manager:v1.4.3
docker.io/longhornio/longhorn-ui:v1.4.3
docker.io/longhornio/support-bundle-kit:v0.0.25
docker.io/rancher/fleet-agent:v0.7.0
docker.io/rancher/fleet:v0.7.0
docker.io/rancher/gitjob:v0.1.54
docker.io/rancher/hardened-calico:v3.25.0-build20230209
docker.io/rancher/hardened-cluster-autoscaler:v1.8.6-build20230406
docker.io/rancher/hardened-cni-plugins:v1.0.1-build20221011
docker.io/rancher/hardened-coredns:v1.10.1-build20230406
docker.io/rancher/hardened-dns-node-cache:1.22.20-build20230406
docker.io/rancher/hardened-etcd:v3.5.7-k3s1-build20230406
docker.io/rancher/hardened-flannel:v0.21.3-build20230308
docker.io/rancher/hardened-k8s-metrics-server:v0.6.2-build20221202
docker.io/rancher/hardened-kubernetes:v1.25.9-rke2r1-build20230412
docker.io/rancher/hardened-multus-cni:v3.9.3-build20230109
docker.io/rancher/harvester-cluster-repo:v1.2.1
docker.io/rancher/harvester-eventrouter:v0.1.1
docker.io/rancher/harvester-load-balancer-webhook:v0.2.1
docker.io/rancher/harvester-load-balancer:v0.2.1
docker.io/rancher/harvester-network-controller:v0.3.4
docker.io/rancher/harvester-network-helper:v0.3.4
docker.io/rancher/harvester-network-webhook:v0.3.4
docker.io/rancher/harvester-node-disk-manager:v0.5.1
docker.io/rancher/harvester-node-manager:v0.1.7
docker.io/rancher/harvester-pcidevices:v0.2.5
docker.io/rancher/harvester-seeder:v0.1.0
docker.io/rancher/harvester-upgrade:v1.2.1
docker.io/rancher/harvester-vm-import-controller:v0.1.5
docker.io/rancher/harvester-webhook:v1.2.1
docker.io/rancher/harvester:v1.2.1
docker.io/rancher/klipper-helm:v0.7.6-build20230223
docker.io/rancher/klipper-lb:v0.4.3
docker.io/rancher/mirrored-library-nginx:1.21.1-alpine
docker.io/rancher/mirrored-sig-storage-snapshot-validation-webhook:v6.2.1
docker.io/rancher/nginx-ingress-controller:nginx-1.6.4-hardened4
docker.io/rancher/rancher-webhook:v0.3.5
docker.io/rancher/rancher:v2.7.5
docker.io/rancher/rke2-runtime:v1.25.9-rke2r1
docker.io/rancher/shell:v0.1.20
docker.io/rancher/shell:v0.1.8
docker.io/rancher/support-bundle-kit:v0.0.28
docker.io/rancher/system-agent-installer-rancher:v2.7.5
docker.io/rancher/system-agent-installer-rke2:v1.25.9-rke2r1
ghcr.io/k8snetworkplumbingwg/whereabouts:v0.6.2-amd64
registry.suse.com/bci/bci-base:15.4
registry.suse.com/suse/sles/15.4/libguestfs-tools:0.54.0-150400.3.19.1
registry.suse.com/suse/sles/15.4/virt-api:0.54.0-150400.3.19.1
registry.suse.com/suse/sles/15.4/virt-controller:0.54.0-150400.3.19.1
registry.suse.com/suse/sles/15.4/virt-handler:0.54.0-150400.3.19.1
registry.suse.com/suse/sles/15.4/virt-launcher:0.54.0-150400.3.19.1
registry.suse.com/suse/sles/15.4/virt-operator:0.54.0-150400.3.19.1
+------------------------------------------------------+
| Images that are going to be REMOVED are listed ABOVE |
+------------------------------------------------------+

```

2. We confirmed those image going to be removed no longer exists after upgrade to the Harvester v1.3.0 master 
  All images list after upgrade to v1.3.0 master
  [master-images2.txt](https://github.com/harvester/harvester/files/14123019/master-images2.txt)


 </details>



### Test Information
* Test Environment: Three nodes harvester on baremetal machines
* Harvester version: `master-70c6cca7-head` (24/01/30)

### Verify Steps
<details><summary>Check the images that are going to be removed after upgrade </summary>

  1. Prepare 3 nodes v1.2.1 Harvester cluster
  2. Upgrade to v1.3.0 master 
  3. After successfully upgrade 
  
  ![image](https://github.com/harvester/harvester/assets/29251855/3b22c006-201a-492c-8582-59e8930dacf0)

  4. Download the [script](https://github.com/harvester/upgrade-helpers/blob/d490e51c76b107c8d5df2df71d0e9b7000b392d6/bin/harv-purge-images.sh) that does image clean-up
  5. Login to `every` Harvester node machines
  6. Execute the command, with the base and target version 
  ```
  ./harv-purge-images.sh v1.2.1 master
  ```
  7. Run the following command to check all available images on the node 
  ```
  ctr -n k8s.io image ls
  ```
  8. Check if any of the image need to be removed exists on the output of step 7. 


</details>
